contrib/urfave/negroni/negroni.go

```diff
@@ -0,0 +1,74 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+// Package negroni provides helper functions for tracing the urfave/negroni package (https://github.com/urfave/negroni).
+package negroni
+
+import (
+	"fmt"
+	"math"
+	"net/http"
+	"strconv"
+
+	"github.com/urfave/negroni"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+// DatadogMiddleware returns middleware that will trace incoming requests.
+type DatadogMiddleware struct {
+	cfg *config
+}
+
+func (m *DatadogMiddleware) ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	opts := []ddtrace.StartSpanOption{
+		tracer.SpanType(ext.SpanTypeWeb),
+		tracer.ServiceName(m.cfg.serviceName),
+		tracer.Tag(ext.HTTPMethod, r.Method),
+		tracer.Tag(ext.HTTPURL, r.URL.Path),
+		tracer.Tag(ext.ResourceName, fmt.Sprintf("%s %s", r.Method, r.URL.Path)),
+		tracer.Measured(),
+	}
+	if !math.IsNaN(m.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, m.cfg.analyticsRate))
+	}
+	if spanctx, err := tracer.Extract(tracer.HTTPHeadersCarrier(r.Header)); err == nil {
+		opts = append(opts, tracer.ChildOf(spanctx))
+	}
+	opts = append(opts, m.cfg.spanOpts...)
+	span, ctx := tracer.StartSpanFromContext(r.Context(), "http.request", opts...)
+	defer span.Finish()
+
+	r = r.WithContext(ctx)
+
+	next(w, r)
+
+	responseWriter := w.(negroni.ResponseWriter)
```

Comment:
I don't think we can assume this is a `negroni.ResponseWriter`. If some other middleware wraps the writer in an incompatible way, this will panic.
Let's *check* and only perform the status stuff if it actually is a `negroni.ResponseWriter`.